plugins/out_pgsql/pgsql.c

```diff
@@ -0,0 +1,238 @@
+                             ctx->db_port,
+                             NULL, NULL,
+                             ctx->db_name,
+                             ctx->db_user,
+                             ctx->db_passwd);
+
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        flb_error("[out_pgsql] failed to connect to host=%s with error: %s",
+                  ctx->db_hostname, PQerrorMessage(ctx->conn));
+
+        pgsql_conf_destroy(ctx);
+
+        return -1;
+    }
+
+    flb_info("[out_pgsql] host=%s port=%s dbname=%s OK",
+              ctx->db_hostname, ctx->db_port, ctx->db_name);
+    flb_output_set_context(ins, ctx);
+
+    temp = PQescapeIdentifier(ctx->conn, ctx->db_table, flb_sds_len(ctx->db_table));
+    flb_sds_destroy(ctx->db_table);
+    ctx->db_table = flb_strdup(temp);
+    flb_free(temp);
+
+    flb_info("[out_pgsql] we check that the table %s exists, if not we create it", ctx->db_table);
+
+    str_len = 62 + flb_sds_len(ctx->db_table);
+
+    query = flb_malloc(str_len);
+    snprintf(query, str_len, "CREATE TABLE IF NOT EXISTS %s (data json);", ctx->db_table);
+    res = PQexec(ctx->conn, query);
+    flb_free(query);
+
+    if(PQresultStatus(res) != PGRES_COMMAND_OK) {
+        flb_error("[out_pgsql] %s", PQerrorMessage(ctx->conn));
+        PQclear(res);
+        return -1;
+    }
+
+    PQclear(res);
+
+    flb_info("[out_pgsql] switching postgresql connection to non-blocking mode");
+    if(PQsetnonblocking(ctx->conn, 1) != 0) {
+        flb_error("[out_pgsql] non-blocking mode not set, not good");
+        return -1;
+    }
+
+    return 0;
+}
+
+static void cb_pgsql_flush(const void *data, size_t bytes,
+                            const char *tag, int tag_len,
+                            struct flb_input_instance *i_ins,
+                            void *out_context,
+                            struct flb_config *config)
+{
+    struct flb_pgsql_config *ctx = out_context;
+    flb_sds_t json;
+    PGresult *res = NULL;
+    int result;
+    char *query = NULL;
+    size_t str_len;
+
+    if(PQconsumeInput(ctx->conn) == 0 && PQisBusy(ctx->conn) == 1) {
+        flb_info("[out_pgsql] Some command may still be running");
+    }
+
+    /*
+      PQreset()
+      This function will close the connection to the server and attempt to reestablish a new connection
+      to the same server, using all the same parameters previously used. This might be useful for error
+      recovery if a working connection is lost.
+     */
+    if(PQstatus(ctx->conn) != CONNECTION_OK) {
+        PQreset(ctx->conn);
+        FLB_OUTPUT_RETURN(FLB_RETRY);
+    }
+
+
+    json = flb_pack_msgpack_to_json_format(data, bytes,
+                                           FLB_PACK_JSON_FORMAT_JSON,
+                                           FLB_PACK_JSON_DATE_DOUBLE,
+                                           ctx->timestamp_key);
+
+    str_len = 115 + flb_sds_len(json);
+
+    query = flb_malloc(str_len);
+    snprintf(query, str_len,
```

Comment:
Check that the memory alloc succeeded

(To be super safe, you could also check the return value of snprintf, but in the Fluent Bit code I've read I haven't generally seen this done..)

Reply:
good catch! a FLB_RETRY should be enough right? I think it's the proper way because if the system run out of memory we need to recheck